sycl/include/CL/sycl/detail/queue_impl.hpp

```diff
@@ -218,6 +230,8 @@ class queue_impl {
 
   bool m_OpenCLInterop = false;
   bool m_HostQueue = false;
+  // Will try to create OOO first.
+  bool m_SupportOOO = true;
```

Comment:
`m_SupportOOO` never gets set to `false`.

Reply:
for createQueue(bool& useOOOQueue) function, the parameter is a reference. m_SupportOOO will be set to false if creation of OOO  failed for the first time.

Reply:
This is really non-intuitive. If you want to update this state in `createQueue`, then you don't need `bool& useOOOQueue` parameter: `createQueue` can access fields of its class explicitly.

Reply:
createQueue() function is set "const" attribute. So field can not be updated in it directly.  
If reference parameter is not preferred, I can add mutable attribute to m_SupportOOO definition. 
which option do you prefer?

Reply:
If `createQueue` is const, it should not try to modify class fields (by reference or by adding `mutable`), unless there is a strong reason to do this. You can make it non-const if you intend to modify something.

Have you considered querying the device capabilities in `queue_impl` ctor and initializing `m_SupportOOO` appropriately? [clGetDeviceInfo](https://www.khronos.org/registry/OpenCL/sdk/1.0/docs/man/xhtml/clGetDeviceInfo.html) with `CL_DEVICE_QUEUE_PROPERTIES` can tell you if OOOQ is supported.

Reply:
Yes, I used this method in test case. Dmitry also give this comments before. But current implementation is ba6sed on the comments in JIRA ( same as commit message).  And major supported device ( CPU and FPGA) support or will support OOO queue. Invoking clGetDeviceInfo will introduce additional API effort in constructor.   Anyway you can make a decision and I will update the patch.

Reply:
I see. In that case, I think you can make `createQueue` non-const and update the state explicitly.